.travis.yml

```diff
@@ -10,35 +10,72 @@ language: minimal
 install: skip
 script: "./dist/linux/travis-ci.sh"
 
+addons:
+  apt:
+    packages:
+      - docker-ce
+
 deploy:
-  - provider: releases
+  - &deploy-docker
+    provider: script
+    skip_cleanup: true
+    script: ./dist/linux/deploy2dockerhub.sh "$IMAGE" "$TRAVIS_COMMIT_MESSAGE"
+    on:
+      repo: 1138-4EB/ghdl
+      branch: master
+  - &deploy-releases
+    provider: releases
```

Comment:
Do we really want to also deploy to docker ?  I am not sure about that.

Reply:
Travis stages are absolutely independent. That means that the images we build in stage 0 (create ghdl/build and ghdl/run images) are not available in the jobs in stage 1 (the matrix expansion and macos). Therefore ghdl/build and ghdl/run images must be saved somewhere. We need some intermediate storage service. This can be S3, dockerhub or any other.

It also acts as an external cache, allowing us to ignore stage 0 for everyday work, and execute it only when platforms change.

However, we can avoid pushing ghdl/ghdl images, which are the ready-to-use images which passed the test successfully. Yet, I think these are the most valuable for end-users. Since we 'need' to set up credentials because of other images in stage 0, I think it is worth reusing them and push these images too.

Reply:
I am a little bit worried about all the network traffic!

I think docker is nice to handle multiple distributions (debian, ubuntu...).
However, I am not convinced about the need of docker images for users.  ghdl is very simple to install, and need other components (like editor).

Reply:
Oh, well, I have not measured it, but I think it is somehow equivalent to the requirements we have now. This is because docker images are [composed of layers](https://docs.docker.com/engine/userguide/storagedriver/imagesandcontainers/#images-and-layers). Therefore, if multiple images have common layers, these need to be downloaded once only.

At now, we build every image for every job in each run. Building means first downloading the base image, then downloading all the dependencies and installing them. E.g., for `stretch-mcode`, the base image has a single 22.49MB layer (e7bb5...) and we then download 40.7MB of archives that we need to install: https://travis-ci.org/tgingold/ghdl/jobs/316767025 Total: 63.19MB + processing time.

With the new approach, we will be downloading the base image and the dependencies all together, already preinstalled in a layer (ghdl/build). Then, we will get a second image (ghdl/run), which is based on it, so most of the content is actually shared and must not be transferred. Last, we build a third image (ghdl/ghdl) based on the latter. This last image also shares all the content with the previous one, except the GHDL tarball. Therefore, when we push the last image to the registry only the last layer needs to be pushed.

Let me show you a practical example with `stretch-mcode`:

![ghdl_layers](https://user-images.githubusercontent.com/6628437/34072538-5e377f72-e289-11e7-9856-9fd73dfdd1b5.gif)

You can see that:

- We need to download two layers, e7bb5... (22.49MB) + 8748e... (70.23MB) to get ghdl/build.
- In order to get ghdl/run, we only need to download layer 3b589... (49.22MB), because layer e7bb5... is already local.
- Last, we only need to download layer f2624... (3.1MB) to get ghdl/ghdl, because the previous two layers, e7bb5... and 3b589... are already local.

In the end, we transferred 22.49 + 70.23 + 49.22 + 3.1 = 145.04MB. In practice, in a normal run we would have downloaded 141.94MB and then uploaded 3.1MB.

Furthermore, I optimized this by using travis' cache system in order to avoid downloading images every time. Long story short, it is stupid because it seems that travis caches are stored exactly in the same place as dockerhub (S3): https://github.com/travis-ci/travis-ci/issues/5358#issuecomment-169856919 . Well, it would make sense if building our images took (tens of) minutes. But we have it done in a minute at most. Anyway, it is straighforward to add it if you want: https://github.com/travis-ci/travis-ci/issues/5358#issuecomment-248915326 , [faster-travis-ci-tests-with-docker-cache](http://atodorov.org/blog/2017/08/07/faster-travis-ci-tests-with-docker-cache/)

---

Yet, we can stil try to optimize it a little: [ghdl/run:stretch-mcode](https://github.com/1138-4EB/ghdl/blob/stages/dist/linux/docker/run/stretch) and [ghdl/build:stretch-mcode](https://github.com/1138-4EB/ghdl/blob/stages/dist/linux/docker/build/stretch), both install gcc in separate layers (3b589... and 8748e..., respectively). We could optimize it installing gcc in its own layer, not with the rest of the dependencies. However, I prefer not to introduce that in this PR, in order to make the transition easier to follow/understand.

---

> However, I am not convinced about the need of docker images for users. ghdl is very simple to install, and need other components (like editor).

Well, we definitely do not need to push ghdl/ghdl images to the registry. But:

- ghdl/run images are already in the registry and we have already built ghdl/ghdl images in order to run tests in them. Thus, pushing ghdl/ghdl images is uploading 3-10MB, the 'nightly' GHDL tarball.
- Any user can have ghdl/ghdl variants built just doing [travis-ci.sh#L10-L16](https://github.com/1138-4EB/ghdl/blob/stages/dist/linux/travis-ci.sh#L10-L16), i.e., using the latest ghdl/run and the lastest tarball available on GitHub releases. We are just avoiding multiple users creating images with the same content but different layer hashes (because they are built independently).
- Use cases in #477.
- There is no conflict at all to use other components (like editor). You can use your usual toolset in the host, by just sharing the folder of the repo with the container. Indeed, you can make the ghdl/ghdl image be an alias of GHDL, so that it is used exactly the same way as if it was installed natively. Of course, that's only one approach. From [#377](https://github.com/tgingold/ghdl/issues/377#issuecomment-352104230):

> Indeed, part of the 'complexity' of docker is due to it's versatility:
> - Daemon mode, how you use it: `docker run -d <image> bash`
> - Batch mode, the example I provided as an alternative to running buildtest.sh locally: `docker run -t -v /$(pwd):/src -w //src <image> bash -c "./dist/linux/buildtest.sh..."`
> - Interactive mode, the demo shown in #477: `docker run -it <image> bash`
> - Any of the previous with GUI support, the second demo shown in #477: `./dist/docker_guiapp.sh -it <image> bash`

- Building GHDL is not difficult, but it is neither straightfordward. It requires several dependencies that a user might not be allowed to install, and we are requiring the user to actually know how programs are compiled and installed (and how to clean temporal files). Making ghdl/ghdl images available makes it equally easy for the user to get mcode, llvm and/or gcc variants. Furthermore, I can have code coverage on windows! Because the stretch-based image with gcc backend will work seamlessly.

- Providing ready-to-use images sets the path for ghdl/ext images, commented in [#477](https://github.com/tgingold/ghdl/issues/477#extend). Having GHDL built with latest vunit and gtkwave versions let's us detect conflicts fast. It also allows us to redirect users to a known conflict-free environment when issues as *VCD works but GHW fails* arise. Note that in this case, a single image is built for each of the tools, all of them based on `ghdl/ghdl:stretch-mcode`, which is my 'official' distribution, as commented below:

Yet, I agree that it might be overkill and not useful to push absolutely all the ghdl/ghdl images to the registry. I think we should pick an "official" distribution, and provide only the variants of it. I'd vote for stretch or buster, because debian slim images are the [smallest](https://hub.docker.com/r/ghdl/ghdl/tags/) (until I can make GHDL run on alpine). Therefore there would be only 3 images available: `ghdl/ghdl:mcode`, `ghdl/ghdl:llvm` and `ghdl/ghdl:gcc`. Optionally, we can add 3 GPL variants: `ghdl/ghdl:mcode-gpl`, `ghdl/ghdl:llvm-gpl` and `ghdl/ghdl:gcc-gpl`.

Reply:
Ok, I was not aware of the optimizing docker image transfer.
Then it would make even more sense not to cache them in travis.